jax/scipy/sparse/linalg.py

```diff
@@ -167,10 +183,306 @@ def cg(A, b, x0=None, *, tol=1e-5, atol=0.0, maxiter=None, M=None):
 
   cg_solve = partial(
       _cg_solve, x0=x0, tol=tol, atol=atol, maxiter=maxiter, M=M)
+
   # real-valued positive-definite linear operators are symmetric
-  real_valued = lambda x: not issubclass(x.dtype.type, np.complexfloating)
+  def real_valued(x):
+    return not issubclass(x.dtype.type, np.complexfloating)
   symmetric = all(map(real_valued, tree_leaves(b)))
   x = lax.custom_linear_solve(
       A, b, solve=cg_solve, transpose_solve=cg_solve, symmetric=symmetric)
   info = None  # TODO(shoyer): return the real iteration count here
   return x, info
+
+
+
+
+def _safe_normalize(x, return_norm=False, thresh=None):
+  """
+  Returns the L2-normalized vector (which can be a pytree) x, and optionally
+  the computed norm. If the computed norm is less than the threshold `thresh`,
+  which by default is the machine precision of x's dtype, it will be
+  taken to be 0, and the normalized x to be the zero vector.
+  """
+  norm = jnp.sqrt(jnp.abs(_vdot_tree(x, x, assume_real=False)))
+  if thresh is None:
+    thresh = jnp.finfo(x.dtype).eps
+  thresh = thresh.astype(norm.dtype)
+
+  normalized_x, norm = lax.cond(
+    norm > thresh,
+    lambda y: (_div(y, norm), norm),
+    lambda y: (tree_map(jnp.zeros_like, y),
+               (thresh*0.).astype(norm.dtype)),  # To get the dtype right
```

Comment:
This should work and is a bit more direct
```suggestion
               jnp.zeros((), dtype=norm.dtype)),  # To get the dtype right
```